src/odemis/acq/align/test/fastem_test.py

```diff
@@ -0,0 +1,185 @@
+
+    def test_optical_autofocus(self):
+        """Run the optical autofocus calibration. Can also be tested with simulator."""
+
+        calibrations = [OPTICAL_AUTOFOCUS]
+
+        # move the stage so that the image is out of focus
+        self.center_position = -30e-6
+        self.focuser.moveAbs({"z": self.center_position}).result()
+
+        # Run auto focus
+        f = fastem.align(self.scanner, self.multibeam, self.descanner, self.mppc, self.stage, self.ccd,
+                         self.beamshift, self.det_rotator, calibrations)
+
+        # foc_pos, foc_lev = f.result(timeout=900)  # TODO make use of the focus level?
+        e = f.result(timeout=900)
+
+        self.assertIsNone(e)  # check no exceptions were returned
+        # check that z stage position is close to good position
+        # Note: This accuracy is dependent on the value chosen for the magnification on the lens.
+        numpy.testing.assert_allclose(self.focuser.position.value["z"], self.good_focus, atol=2e-6)
+
+    def test_image_translation_prealign(self):
+        """Run the image translation prealing calibration. Can also be tested with simulator.
+        It calibrates the descanner offset."""
+
+        calibrations = [IMAGE_TRANSLATION_PREALIGN]
+
+        # get current descanner offset
+        descan_offset_cur = self.descanner.scanOffset.value
+
+        # Run image translation pre-align
+        f = fastem.align(self.scanner, self.multibeam, self.descanner, self.mppc, self.stage, self.ccd,
+                         self.beamshift, self.det_rotator, calibrations)
+
+        e = f.result(timeout=900)
+
+        self.assertIsNone(e)  # check no exceptions were returned
+
+        # get the calibrated descanner offset
+        descan_offset_calib = self.descanner.scanOffset.value
+
+        # check the calibrated descan offset is different from the previous offset
+        self.assertNotEqual(descan_offset_cur, descan_offset_calib)
+
+    def test_progress(self):
+        """Check if some progress is reported during the optical autofocus calibration."""
+
+        self.updates = 0  # updated in callback on_progress_update
+
+        calibrations = [OPTICAL_AUTOFOCUS, IMAGE_TRANSLATION_PREALIGN]
+        f = fastem.align(self.scanner, self.multibeam, self.descanner, self.mppc, self.stage, self.ccd,
+                         self.beamshift, self.det_rotator, calibrations)
+
+        f.add_update_callback(self.on_progress_update)  # callback executed every time f.set_progress is called
+        f.add_done_callback(self.on_done)  # callback executed when f.set_result is called (via bindFuture)
+
+        e = f.result()
+
+        self.assertIsNone(e)  # check no exceptions were returned
+        self.assertTrue(self.done)
+        # at least one update per calibration plus once at start of calibration, plus once at end of calibration
+        self.assertGreaterEqual(self.updates, 4)  # at least one update per
```

Comment:
why use the `assertGreaterEqual` and not `assertEqual`, can't you be sure that there are exactly 4 updates?

Reply:
No I cannot. The problem with these set_progress calls is that it is not only called in the calibration manager explicitly, but also when the future starts and ends I found out. So even if I never call set_progress here in the manager, it is still called. In other test cases I only find assertGreaterEqual so I would go for this to avoid failing tests. As long as progress is found we are happy I think.